File: typescript/packages/bench/src/repositories/partRepository.ts
/** Read side over the parts catalogue. */

import type { Repository } from "@atelier/core";
import type { Part } from "@atelier/core";

/** Read side over the parts catalogue. */
export class PartRepository implements Repository<number, Part> {
    constructor(private readonly records: readonly Part[] = []) {}

    find(id: number): Part | undefined {
        return this.records.find((record) => record.id === id);
    }

    all(): readonly Part[] {
        return this.records;
    }

    count(): number {
        return this.records.length;
    }

    /** Catalogue lookup by stock-keeping unit. */
    bySku(sku: string): Part | undefined {
        return this.records.find((part) => part.sku === sku);
    }
}

File: rust/crates/atelier-bench/src/repositories/part_repository.rs
//! Read side over the parts catalogue.

use atelier_core::Part;
use atelier_core::contracts::repository::Repository;

/// Read side over the parts catalogue.
#[derive(Debug, Clone, Default)]
pub struct PartRepository {
    records: Vec<Part>,
}

impl PartRepository {
    pub fn new(records: Vec<Part>) -> Self {
        Self { records }
    }

    /// Catalogue lookup by stock-keeping unit.
    pub fn by_sku(&self, sku: &str) -> Option<&Part> {
        self.records.iter().find(|part| part.sku == sku)
    }
}

impl Repository for PartRepository {
    type Id = u32;
    type Record = Part;

    fn find(&self, id: Self::Id) -> Option<&Self::Record> {
        self.records.iter().find(|record| record.id == id)
    }

    fn all(&self) -> &[Self::Record] {
        &self.records
    }
}
